pkg/file/xar.go

```diff
@@ -181,43 +190,69 @@ func ExtractXARMetadata(r io.Reader) (*InstallerMetadata, error) {
 	heapOffset := xarHeaderSize + hdr.CompressedSize
 	for _, f := range root.TOC.Files {
 		if f.Name == "Distribution" {
-			var fileReader io.Reader
-			heapReader := io.NewSectionReader(rr, heapOffset, int64(len(b))-heapOffset)
-			fileReader = io.NewSectionReader(heapReader, f.Data.Offset, f.Data.Length)
-
-			// the distribution file can be compressed differently than the TOC, the
-			// actual compression is specified in the Encoding.Style field.
-			if strings.Contains(f.Data.Encoding.Style, "x-gzip") {
-				// despite the name, x-gzip fails to decode with the gzip package
-				// (invalid header), but it works with zlib.
-				zr, err := zlib.NewReader(fileReader)
-				if err != nil {
-					return nil, fmt.Errorf("create zlib reader: %w", err)
-				}
-				defer zr.Close()
-				fileReader = zr
-			} else if strings.Contains(f.Data.Encoding.Style, "x-bzip2") {
-				fileReader = bzip2.NewReader(fileReader)
-			}
-			// TODO: what other compression methods are supported?
-
-			contents, err := io.ReadAll(fileReader)
+			contents, err := readCompressedFile(rr, heapOffset, int64(len(b)), f)
 			if err != nil {
-				return nil, fmt.Errorf("reading Distribution file: %w", err)
+				return nil, err
 			}
 
 			meta, err := parseDistributionFile(contents)
 			if err != nil {
 				return nil, fmt.Errorf("parsing Distribution file: %w", err)
 			}
 			meta.SHASum = h.Sum(nil)
-			return meta, err
+			return meta, nil
+		}
+	}
+
+	// If Distribution archive was not found, we look for a top-level PackageInfo archive
+	// Unofficial specs: http://s.sudre.free.fr/Stuff/Ivanhoe/FLAT.html
+	for _, f := range root.TOC.Files {
+		if f.Name == "PackageInfo" {
```

Comment:
Instead of looping twice over the files, how about keeping the package info in a separate metadata variable and after the first (only) loop return it if distro was not found?

Reply:
Good suggestion. I made the fix.